lua/neo-tree/sources/common/preview.lua

```diff
@@ -0,0 +1,209 @@
+  vim.api.nvim_win_set_option(self.winid, "foldenable", self.truth.options.foldenable)
+  vim.api.nvim_win_set_var(self.winid, "neo_tree_preview", 0)
+end
+
+---Subscribe to event and add it to the preview event list.
+--@param source string? Name of the source to add the event to. Will use `events.subscribe` if nil.
+--@param event table Event to subscribe to.
+function Preview:subscribe(source, event)
+  if source == nil then
+    events.subscribe(event)
+  else
+    manager.subscribe(source, event)
+  end
+  self.events = self.events or {}
+  table.insert(self.events, { source = source, event = event })
+end
+
+---Unsubscribe to all events in the preview event list.
+function Preview:unsubscribe()
+  for _, event in ipairs(self.events) do
+    if event.source == nil then
+      events.unsubscribe(event.event)
+    else
+      manager.unsubscribe(event.source, event.event)
+    end
+  end
+  self.events = {}
+end
+
+---Finds the appropriate window and updates the preview accordingly.
+---@param state table The state of the source.
+function Preview:findWindow(state)
+  local winid, is_neo_tree_window = utils.get_appropriate_window(state)
+  if winid == self.winid then
+    return
+  end
+
+  if self.active then
+    self:revert()
+    self.winid, self.is_neo_tree_window = winid, is_neo_tree_window
+    self:preview()
+  else
+    self.winid, self.is_neo_tree_window = winid, is_neo_tree_window
+    self.bufnr = vim.api.nvim_win_get_buf(self.winid)
+  end
+end
+
+---Activates the preview, but does not populate the preview window,
+function Preview:activate()
+  if self.active then
+    return
+  end
+  self.truth = {
+    bufnr = self.bufnr,
+    view = vim.api.nvim_win_call(self.winid, vim.fn.winsaveview),
+    options = {
+      bufhidden = vim.api.nvim_buf_get_option(self.bufnr, "bufhidden"),
+      foldenable = vim.api.nvim_win_get_option(self.winid, "foldenable"),
+    },
+  }
+  vim.api.nvim_buf_set_option(self.bufnr, "bufhidden", "hide")
+  vim.api.nvim_win_set_option(self.winid, "foldenable", false)
+  self.active = true
+  vim.api.nvim_win_set_var(self.winid, "neo_tree_preview", 1)
+end
+
+---Set the buffer in the preview window without executing BufEnter or BufWinEnter autocommands.
+--@param bufnr number The buffer number of the buffer to set.
+function Preview:setBuffer(bufnr)
+  local eventignore = vim.opt.eventignore
+  vim.opt.eventignore:append("BufEnter,BufWinEnter")
+  vim.api.nvim_win_set_buf(self.winid, bufnr)
+  vim.opt.eventignore = eventignore
+end
+
+---Move the cursor to the previewed position and center the screen.
+function Preview:reveal()
+  local pos = self.start_pos or self.end_pos
+  if not self.active or not self.winid or not pos then
+    return
+  end
+  vim.api.nvim_win_set_cursor(self.winid, { (pos[1] or 0) + 1, pos[2] or 0 })
+  vim.api.nvim_win_call(self.winid, function()
+    vim.cmd("normal! zz")
+  end)
+end
+
+---Highlight the previewed range
+function Preview:highlight()
```

Comment:
What is this code supposed to do? I don't notice any special highlights when I use the preview, but the highlight group does have settings which are different from Normal.

Reply:
It doesn't seem to work for me. It also does not jump to the correct line for the focused diagnostic, is it supposed to do that?

![image](https://user-images.githubusercontent.com/5160605/184933328-a6c716b1-8a3a-4bae-b328-1f205f8c0056.png)

Reply:
Ah, you're probably just not using the `native-open-and-preview` branch I set up to test this. The preview and the open commands look to `node.extra.position` and `node.extra.end_position` to determine what to highlight and where to open. The main branch uses `node.extra.diag_struct` for its own open commands and does not set these fields.

Reply:
Ahhh, OK. I think this is good to merge now.

Reply:
Alright, sounds good. I might revisit the highlight logic at a later date, but for now it's fine.